lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp

```diff
@@ -1523,6 +1599,194 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
+  // avoids re-use of the phi arg by the SSAUpdater for a subsequent redundant
+  // load thereby avoiding multiple consuming uses of the phi arg.
+  if (auto *Phi = dyn_cast<SILPhiArgument>(ForwardingValue)) {
+    NewPhiArgs.insert(cast<SILPhiArgument>(ForwardingValue));
+    ForwardingValue = createNewCopy(
+        &*getInsertAfterPoint(ForwardingValue).getValue(), ForwardingValue);
+    // Mark as forwarded
+    ForwardedValues.insert(ForwardingValue);
+    // If the forwarding value is not in the same loop as the redundant load,
+    // create a copy before the load
+    if (needsAdditionalCopyBeforeRedundantLoad(RedundantLoad,
+                                               ForwardingValue)) {
+      ForwardingValue = createNewCopy(RedundantLoad, ForwardingValue);
+    }
+    // Mark as forwarded
+    ForwardedValues.insert(ForwardingValue);
+    return ForwardingValue;
+  }
+
+  // If the forwarded value is a StructInst or a TupleInst, go over all its
+  // operands which are copy_values and mark them as forwarded.
+  // Also in this case, the forwarding value will never be reused, so we don't
+  // have to create a copy.
+  //
+  // The forwarding value is a StructInst when a dominating value of the same
+  // memory is not available, but dominating values of all its fields are
+  // available. Then, LSValue::reduce will create a struct from values of all
+  // its available fields.
+  // Similarly, the forwarding value is a TupleInst, when values of all this
+  // elements are available.
+  if (isa<StructInst>(ForwardingValue) || isa<TupleInst>(ForwardingValue)) {
+    auto *Inst = cast<SingleValueInstruction>(ForwardingValue);
+    for (auto &Op : Inst->getAllOperands()) {
+      auto *OpVal = cast<CopyValueInst>(Op.get());
+      assert(NewCopies.find(OpVal) != NewCopies.end());
+      ForwardedValues.insert(OpVal);
+    }
+    return ForwardingValue;
+  }
+
+  assert(0 && "Unhandled case in RLE");
+  return ForwardingValue;
+}
+
+// Given a newly created value 'Val', insert destroy_value at its frontiers
+// when not already found
+void RLEContext::fixupLifetimeOfValue(SILValue Val) {
+  SmallVector<SILInstruction *, 4> UserSet;
+  for (auto *Use : Val->getUses()) {
+    if (auto *Borrow = dyn_cast<BeginBorrowInst>(Use->getUser())) {
+      auto EndBorrows = Borrow->getEndBorrows();
+      UserSet.append(EndBorrows.begin(), EndBorrows.end());
+    } else {
+      UserSet.push_back(Use->getUser());
+    }
+  }
+  PointerUnion<SILInstruction *, SILArgument *> Def;
+  if (auto *Copy = dyn_cast<CopyValueInst>(Val)) {
+    Def = Copy;
+  } else {
+    Def = cast<SILPhiArgument>(Val);
+  }
+
+  ValueLifetimeAnalysis Vla(Def, UserSet);
+  ValueLifetimeAnalysis::Frontier ValFrontier;
+  Vla.computeFrontier(ValFrontier, ValueLifetimeAnalysis::DontModifyCFG);
+  // Go over all the frontiers and insert a destroy of the copy if a
+  // lifetime ending instruction was not found at the frontier
+  for (auto *End : ValFrontier) {
```

Comment:
@gottesmm I could not use the `findJointPostDominatingSet` utility here. I have some copies in RLE that will not have any consuming use. And I need to figure out the correct BasicBlock::iterator to insert the destroy_value.